docs/adr/0003-workspace-blob-caching.md

```diff
@@ -0,0 +1,56 @@
+# 3. Workspace BLOB Caching
+
+* Status: [ **proposed** | rejected | accepted | deprecated ]
+* Date: 2023-09-20 
+* Authors: @chanwit
+* Deciders: TBD 
+
+## Context
+
+The TF-Controller currently faces challenges related to the deletion of Terraform resources.
+These problems span across three categories:
+
+1. Single object deletion,
+2. Resources with dependencies deletion, and
+3. Namespace deletion.
+
+These problems must be fixed in the above order as (2) and (3) require single object deletion to be resolved first.
+
+Deleting a single TF object can sometimes be obstructed because it's tied to other resources like Source objects, Secrets, and ConfigMaps. If we try to remove it without deleting these resources, the TF object gets stuck in an inconsistent state, making it harder for users to manage their infrastructure smoothly.
```

Comment:
Based on the solution given below I would expect the problem to be that a Terraform object can't be finalized if resources it depends on are deleted _before_ it is. The description here suggests vice versa: that deleting a Terraform object before the things it relies on leads to the problem. Have I misunderstood?